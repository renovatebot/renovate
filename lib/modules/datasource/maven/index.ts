import is from '@sindresorhus/is';
import { DateTime } from 'luxon';
import type { XmlDocument } from 'xmldoc';
import { logger } from '../../../logger';
import * as packageCache from '../../../util/cache/package';
import * as p from '../../../util/promises';
import { newlineRegex, regEx } from '../../../util/regex';
import { ensureTrailingSlash } from '../../../util/url';
import mavenVersion from '../../versioning/maven';
import * as mavenVersioning from '../../versioning/maven';
import { compare } from '../../versioning/maven/compare';
import { Datasource } from '../datasource';
import type {
  GetReleasesConfig,
  RegistryStrategy,
  Release,
  ReleaseResult,
} from '../types';
import { MAVEN_REPO } from './common';
import type { MavenDependency, ReleaseMap } from './types';
import {
  checkResource,
  createUrlForDependencyPom,
  downloadHttpProtocol,
  downloadMavenXml,
  getDependencyInfo,
  getDependencyParts,
  getMavenUrl,
} from './util';

function getLatestSuitableVersion(releases: Release[]): string | null {
  // istanbul ignore if
  if (!releases?.length) {
    return null;
  }
  const allVersions = releases.map(({ version }) => version);
  const stableVersions = allVersions.filter((x) => mavenVersion.isStable(x));
  const versions = stableVersions.length ? stableVersions : allVersions;
  return versions.reduce((latestVersion, version) =>
    compare(version, latestVersion) === 1
      ? version
      : /* istanbul ignore next: hard to test */ latestVersion
  );
}

function extractVersions(metadata: XmlDocument): string[] {
  const versions = metadata.descendantWithPath('versioning.versions');
  const elements = versions?.childrenNamed('version');
  if (!elements) {
    return [];
  }
  return elements.map((el) => el.val);
}

const mavenCentralHtmlVersionRegex = regEx(
  '^<a href="(?<version>[^"]+)\\/" title="(?:[^"]+)\\/">(?:[^"]+)\\/<\\/a>\\s+(?<releaseTimestamp>\\d\\d\\d\\d-\\d\\d-\\d\\d \\d\\d:\\d\\d)\\s+-$',
  'i'
);

export const defaultRegistryUrls = [MAVEN_REPO];

export class MavenDatasource extends Datasource {
  static id = 'maven';

  override readonly defaultRegistryUrls = defaultRegistryUrls;

  override readonly defaultVersioning: string = mavenVersioning.id;

  override readonly registryStrategy: RegistryStrategy = 'merge';

  constructor(id = MavenDatasource.id) {
    super(id);
  }

  async fetchReleasesFromMetadata(
    dependency: MavenDependency,
    repoUrl: string
  ): Promise<ReleaseMap> {
    const metadataUrl = getMavenUrl(dependency, repoUrl, 'maven-metadata.xml');

    const cacheNamespace = 'datasource-maven:metadata-xml';
    const cacheKey = metadataUrl.toString();
    const cachedVersions = await packageCache.get<ReleaseMap>(
      cacheNamespace,
      cacheKey
    );
    /* istanbul ignore if */
    if (cachedVersions) {
      return cachedVersions;
    }

    const { isCacheable, xml: mavenMetadata } = await downloadMavenXml(
      this.http,
      metadataUrl
    );
    if (!mavenMetadata) {
      return {};
    }

    const versions = extractVersions(mavenMetadata);
    const releaseMap = versions.reduce(
      (acc, version) => ({ ...acc, [version]: null }),
      {}
    );
    if (isCacheable) {
      await packageCache.set(cacheNamespace, cacheKey, releaseMap, 30);
    }
    return releaseMap;
  }

  async addReleasesFromIndexPage(
    inputReleaseMap: ReleaseMap,
    dependency: MavenDependency,
    repoUrl: string
  ): Promise<ReleaseMap> {
    const cacheNs = 'datasource-maven:index-html-releases';
    const cacheKey = `${repoUrl}${dependency.dependencyUrl}`;
    let workingReleaseMap = await packageCache.get<ReleaseMap>(
      cacheNs,
      cacheKey
    );
    if (!workingReleaseMap) {
      workingReleaseMap = {};
      let retryEarlier = false;
      try {
        if (repoUrl.startsWith(MAVEN_REPO)) {
          const indexUrl = getMavenUrl(dependency, repoUrl, 'index.html');
          const res = await downloadHttpProtocol(this.http, indexUrl);
          const { body = '' } = res;
          for (const line of body.split(newlineRegex)) {
            const match = line.trim().match(mavenCentralHtmlVersionRegex);
            if (match) {
              const { version, releaseTimestamp: timestamp } =
                match?.groups ?? /* istanbul ignore next: hard to test */ {};
              if (version && timestamp) {
                const date = DateTime.fromFormat(
                  timestamp,
                  'yyyy-MM-dd HH:mm',
                  {
                    zone: 'UTC',
                  }
                );
                if (date.isValid) {
                  const releaseTimestamp = date.toISO();
                  workingReleaseMap[version] = { version, releaseTimestamp };
                }
              }
            }
          }
        }
      } catch (err) /* istanbul ignore next */ {
        retryEarlier = true;
        logger.debug(
          { dependency, err },
          'Failed to get releases from index.html'
        );
      }
      const cacheTTL = retryEarlier
        ? /* istanbul ignore next: hard to test */ 60
        : 24 * 60;
      await packageCache.set(cacheNs, cacheKey, workingReleaseMap, cacheTTL);
    }

    const releaseMap = { ...inputReleaseMap };
    for (const version of Object.keys(releaseMap)) {
      releaseMap[version] ||= workingReleaseMap[version] ?? null;
    }

    return releaseMap;
  }

  /**
   *
   * Double-check releases using HEAD request and
   * attach timestamps obtained from `Last-Modified` header.
   *
   * Example input:
   *
   * {
   *   '1.0.0': {
   *     version: '1.0.0',
   *     releaseTimestamp: '2020-01-01T01:00:00.000Z',
   *   },
   *   '1.0.1': null,
   * }
   *
   * Example output:
   *
   * {
   *   '1.0.0': {
   *     version: '1.0.0',
   *     releaseTimestamp: '2020-01-01T01:00:00.000Z',
   *   },
   *   '1.0.1': {
   *     version: '1.0.1',
   *     releaseTimestamp: '2021-01-01T01:00:00.000Z',
   *   }
   * }
   *
   * It should validate `1.0.0` with HEAD request, but leave `1.0.1` intact.
   *
   */
  async addReleasesUsingHeadRequests(
    inputReleaseMap: ReleaseMap,
    dependency: MavenDependency,
    repoUrl: string
  ): Promise<ReleaseMap> {
    const releaseMap = { ...inputReleaseMap };

    if (process.env.RENOVATE_EXPERIMENTAL_NO_MAVEN_POM_CHECK) {
      return releaseMap;
    }

    const cacheNs = 'datasource-maven:head-requests';
    const cacheTimeoutNs = 'datasource-maven:head-requests-timeout';
    const cacheKey = `${repoUrl}${dependency.dependencyUrl}`;

    // Store cache validity as the separate flag.
    // This allows both cache updating and resetting.
    //
    // Even if new version is being released each 10 minutes,
    // we still want to reset the whole cache after 24 hours.
    const isCacheValid = await packageCache.get<true>(cacheTimeoutNs, cacheKey);

    let cachedReleaseMap: ReleaseMap = {};
    // istanbul ignore if
    if (isCacheValid) {
      const cache = await packageCache.get<ReleaseMap>(cacheNs, cacheKey);
      if (cache) {
        cachedReleaseMap = cache;
      }
    }

    // List versions to check with HEAD request
    const freshVersions = Object.entries(releaseMap)
      .filter(([version, release]) => {
        // Release is present in maven-metadata.xml,
        // but haven't been validated yet
        const isValidatedAtPreviousSteps = release !== null;

        // Release was validated and cached with HEAD request during previous run
        const isValidatedHere = !is.undefined(cachedReleaseMap[version]);

        // Select only valid releases not yet verified with HEAD request
        return !isValidatedAtPreviousSteps && !isValidatedHere;
      })
      .map(([k]) => k);

    // Update cached data with freshly discovered versions
    if (freshVersions.length) {
      const queue = freshVersions.map((version) => async (): Promise<void> => {
        const pomUrl = await createUrlForDependencyPom(
          this.http,
          version,
          dependency,
          repoUrl
        );
        const artifactUrl = getMavenUrl(dependency, repoUrl, pomUrl);
        const release: Release = { version };

        const res = await checkResource(this.http, artifactUrl);

        if (is.date(res)) {
          release.releaseTimestamp = res.toISOString();
        }

        cachedReleaseMap[version] =
          res !== 'not-found' && res !== 'error' ? release : null;
      });

      await p.all(queue);

      if (!isCacheValid) {
        // Store new TTL flag for 24 hours if the previous one is invalidated
        await packageCache.set(cacheTimeoutNs, cacheKey, 'long', 24 * 60);
      }

      // Store updated cache object
      await packageCache.set(cacheNs, cacheKey, cachedReleaseMap, 24 * 60);
    }

    // Filter releases with the versions validated via HEAD request
    for (const version of Object.keys(releaseMap)) {
      releaseMap[version] = cachedReleaseMap[version] ?? null;
    }
    return releaseMap;
  }

  getReleasesFromMap(releaseMap: ReleaseMap): Release[] {
    const releases = Object.values(releaseMap).filter(is.truthy);
    if (releases.length) {
      return releases;
    }
    return Object.keys(releaseMap).map((version) => ({ version }));
  }

  async getReleases({
    packageName,
    registryUrl,
  }: GetReleasesConfig): Promise<ReleaseResult | null> {
    // istanbul ignore if
    if (!registryUrl) {
      return null;
    }

    const dependency = getDependencyParts(packageName);
    const repoUrl = ensureTrailingSlash(registryUrl);

    logger.debug(`Looking up ${dependency.display} in repository ${repoUrl}`);

    let releaseMap = await this.fetchReleasesFromMetadata(dependency, repoUrl);
    releaseMap = await this.addReleasesFromIndexPage(
      releaseMap,
      dependency,
      repoUrl
    );
    releaseMap = await this.addReleasesUsingHeadRequests(
      releaseMap,
      dependency,
      repoUrl
    );
    const releases = this.getReleasesFromMap(releaseMap);
    if (!releases?.length) {
      return null;
    }

    logger.debug(
      `Found ${releases.length} new releases for ${dependency.display} in repository ${repoUrl}`
    );

    const latestSuitableVersion = getLatestSuitableVersion(releases);
    const dependencyInfo =
      latestSuitableVersion &&
      (await getDependencyInfo(
        this.http,
        dependency,
        repoUrl,
        latestSuitableVersion
      ));

    return { ...dependency, ...dependencyInfo, releases };
  }
}
