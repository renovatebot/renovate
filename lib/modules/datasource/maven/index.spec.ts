import { ReleaseResult, getPkgReleases } from '..';
import { Fixtures } from '../../../../test/fixtures';
import * as httpMock from '../../../../test/http-mock';
import { EXTERNAL_HOST_ERROR } from '../../../constants/error-messages';
import * as hostRules from '../../../util/host-rules';
import { id as versioning } from '../../versioning/maven';
import { MavenDatasource } from '.';

const datasource = MavenDatasource.id;

const baseUrl = 'https://repo.maven.apache.org/maven2';
const baseUrlCustom = 'https://custom.registry.renovatebot.com';

interface SnapshotOpts {
  version: string;
  jarStatus?: number;
  meta?: string;
}

interface MockOpts {
  dep?: string;
  base?: string;
  meta?: string | null;
  pom?: string | null;
  latest?: string;
  jars?: Record<string, number> | null;
  snapshots?: SnapshotOpts[] | null;
  html?: string | null;
}

function mockGenericPackage(opts: MockOpts = {}) {
  const {
    dep = 'org.example:package',
    base = baseUrl,
    latest = '2.0.0',
    html,
  } = opts;
  const meta =
    opts.meta === undefined ? Fixtures.get('metadata.xml') : opts.meta;
  const pom = opts.pom === undefined ? Fixtures.get('pom.xml') : opts.pom;
  const jars =
    opts.jars === undefined
      ? {
          '0.0.1': 200,
          '1.0.0': 200,
          '1.0.1': 404,
          '1.0.2': 500,
          '2.0.0': 200,
        }
      : opts.jars;

  const snapshots =
    opts.snapshots === undefined
      ? [
          {
            version: '1.0.3-SNAPSHOT',
            meta: Fixtures.get('metadata-snapshot-version.xml'),
            jarStatus: 200,
          },
          {
            version: '1.0.4-SNAPSHOT',
            meta: Fixtures.get('metadata-snapshot-version-invalid.xml'),
          },
          {
            version: '1.0.5-SNAPSHOT',
          },
        ]
      : opts.snapshots;

  const scope = httpMock.scope(base);

  const [group, artifact] = dep.split(':');
  const packagePath = `${group.replace(/\./g, '/')}/${artifact}`;

  if (meta) {
    scope.get(`/${packagePath}/maven-metadata.xml`).reply(200, meta);
  }

  if (html) {
    scope.get(`/${packagePath}/index.html`).reply(200, html);
  } else if (html === null) {
    scope.get(`/${packagePath}/index.html`).reply(404);
  }

  if (pom) {
    if (latest.endsWith('-SNAPSHOT')) {
      const [major, minor, patch] = latest
        .replace('-SNAPSHOT', '')
        .split('.')
        .map((x) => parseInt(x, 10))
        .map((x) => (x < 10 ? `0${x}` : `${x}`));
      scope
        .get(
          `/${packagePath}/${latest}/${artifact}-${latest.replace(
            '-SNAPSHOT',
            ''
          )}-20200101.${major}${minor}${patch}-${parseInt(patch, 10)}.pom`
        )
        .reply(200, pom);
    } else {
      scope
        .get(`/${packagePath}/${latest}/${artifact}-${latest}.pom`)
        .reply(200, pom);
    }
  }

  if (jars) {
    Object.entries(jars).forEach(([version, status]) => {
      const [major, minor, patch] = version
        .replace('-SNAPSHOT', '')
        .split('.')
        .map((x) => parseInt(x, 10))
        .map((x) => (x < 10 ? `0${x}` : `${x}`));
      const timestamp = `2020-01-01T${major}:${minor}:${patch}.000Z`;
      const headers: httpMock.ReplyHeaders = version.startsWith('0.')
        ? {}
        : { 'Last-Modified': timestamp };
      scope
        .head(`/${packagePath}/${version}/${artifact}-${version}.pom`)
        .reply(status, '', headers);
    });
  }

  if (snapshots) {
    snapshots.forEach((snapshot) => {
      if (snapshot.meta) {
        scope
          .get(`/${packagePath}/${snapshot.version}/maven-metadata.xml`)
          .reply(200, snapshot.meta);
      } else {
        scope
          .get(`/${packagePath}/${snapshot.version}/maven-metadata.xml`)
          .reply(404, '');
      }

      if (snapshot.jarStatus) {
        const [major, minor, patch] = snapshot.version
          .replace('-SNAPSHOT', '')
          .split('.')
          .map((x) => parseInt(x, 10))
          .map((x) => (x < 10 ? `0${x}` : `${x}`));
        const timestamp = `2020-01-01T${major}:${minor}:${patch}.000Z`;
        scope
          .head(
            `/${packagePath}/${
              snapshot.version
            }/${artifact}-${snapshot.version.replace(
              '-SNAPSHOT',
              ''
            )}-20200101.${major}${minor}${patch}-${parseInt(patch, 10)}.pom`
          )
          .reply(snapshot.jarStatus, '', { 'Last-Modified': timestamp });
      } else {
        scope
          .head(
            `/${packagePath}/${snapshot.version}/${artifact}-${snapshot.version}.pom`
          )
          .reply(404, '');
      }
    });
  }
}

function get(
  depName = 'org.example:package',
  ...registryUrls: string[]
): Promise<ReleaseResult | null> {
  const conf = { versioning, datasource, depName };
  return getPkgReleases(registryUrls ? { ...conf, registryUrls } : conf);
}

describe('modules/datasource/maven/index', () => {
  beforeEach(() => {
    hostRules.add({
      hostType: datasource,
      matchHost: 'custom.registry.renovatebot.com',
      token: '123test',
    });
    jest.resetAllMocks();
  });

  afterEach(() => {
    hostRules.clear();
    delete process.env.RENOVATE_EXPERIMENTAL_NO_MAVEN_POM_CHECK;
  });

  it('returns null when metadata is not found', async () => {
    httpMock
      .scope(baseUrl)
      .get('/org/example/package/index.html')
      .reply(404)
      .get('/org/example/package/maven-metadata.xml')
      .reply(404);

    const res = await get();

    expect(res).toBeNull();
  });

  it('returns releases', async () => {
    mockGenericPackage({ html: null });

    const res = await get();

    expect(res).toMatchSnapshot();
  });

  it('returns releases when only snapshot', async () => {
    const meta = Fixtures.get('metadata-snapshot-version.xml');
    mockGenericPackage({
      meta: Fixtures.get('metadata-snapshot-only.xml'),
      jars: null,
      html: null,
      latest: '1.0.3-SNAPSHOT',
      snapshots: [
        {
          version: '1.0.3-SNAPSHOT',
          meta,
          jarStatus: 200,
        },
      ],
    });
    httpMock
      .scope(baseUrl)
      .get('/org/example/package/1.0.3-SNAPSHOT/maven-metadata.xml')
      .reply(200, meta);

    const res = await get();

    expect(res).toEqual({
      display: 'org.example:package',
      group: 'org.example',
      homepage: 'https://package.example.org/about',
      name: 'package',
      registryUrl: 'https://repo.maven.apache.org/maven2',
      releases: [
        {
          releaseTimestamp: '2020-01-01T01:00:03.000Z',
          version: '1.0.3-SNAPSHOT',
        },
      ],
    });
  });

  it('returns html-based releases', async () => {
    process.env.RENOVATE_EXPERIMENTAL_NO_MAVEN_POM_CHECK = 'true';

    mockGenericPackage({
      latest: '2.0.0',
      jars: null,
      html: Fixtures.get('index.html'),
      meta: Fixtures.get('index.xml'),
      snapshots: null,
    });

    const res = await get();

    expect(res).toEqual({
      display: 'org.example:package',
      group: 'org.example',
      homepage: 'https://package.example.org/about',
      name: 'package',
      registryUrl: 'https://repo.maven.apache.org/maven2',
      releases: [
        { version: '1.0.0', releaseTimestamp: '2021-02-22T14:43:00.000Z' },
        { version: '1.0.1', releaseTimestamp: '2021-04-12T15:51:00.000Z' },
        { version: '1.0.2', releaseTimestamp: '2021-06-16T12:47:00.000Z' },
        { version: '2.0.0', releaseTimestamp: '2021-06-18T16:24:00.000Z' },
      ],
    });
  });

  it('returns releases from custom repository', async () => {
    mockGenericPackage({ base: baseUrlCustom });

    const res = await get('org.example:package', baseUrlCustom);

    expect(res).toMatchSnapshot();
  });

  it('collects releases from all registry urls', async () => {
    mockGenericPackage({ html: null });
    mockGenericPackage({
      base: baseUrlCustom,
      meta: Fixtures.get('metadata-extra.xml'),
      latest: '3.0.0',
      jars: { '3.0.0': 200 },
      snapshots: [],
    });

    const res = await get('org.example:package', baseUrl, baseUrlCustom);

    expect(res?.releases).toMatchObject([
      { version: '0.0.1' },
      { version: '1.0.0' },
      { version: '1.0.3-SNAPSHOT' },
      { version: '2.0.0' },
      { version: '3.0.0' },
    ]);
  });

  it('falls back to next registry url', async () => {
    mockGenericPackage({ html: null });
    httpMock
      .scope('https://failed_repo')
      .get('/org/example/package/maven-metadata.xml')
      .reply(404);
    httpMock
      .scope('https://unauthorized_repo')
      .get('/org/example/package/maven-metadata.xml')
      .reply(403);
    httpMock
      .scope('https://empty_repo')
      .get('/org/example/package/maven-metadata.xml')
      .reply(200, 'non-sense');
    httpMock
      .scope('https://unknown_error')
      .get('/org/example/package/maven-metadata.xml')
      .replyWithError('unknown');

    const res = await get(
      'org.example:package',
      'https://failed_repo/',
      'https://unauthorized_repo/',
      'https://empty_repo',
      'https://unknown_error',
      baseUrl
    );

    expect(res).toMatchSnapshot();
  });

  it('throws EXTERNAL_HOST_ERROR for 50x', async () => {
    httpMock
      .scope(baseUrl)
      .get('/org/example/package/maven-metadata.xml')
      .reply(503);

    await expect(get()).rejects.toThrow(EXTERNAL_HOST_ERROR);
  });

  it('ignores unsupported protocols', async () => {
    const base = baseUrl.replace('https', 'http');
    mockGenericPackage({ base });

    const res = await get(
      'org.example:package',
      'ftp://protocol_error_repo',
      base
    );

    expect(res?.releases).toMatchSnapshot();
  });

  it('skips registry with invalid metadata structure', async () => {
    mockGenericPackage({ html: null });
    httpMock
      .scope('https://invalid_metadata_repo')
      .get('/org/example/package/maven-metadata.xml')
      .reply(200, Fixtures.get('metadata-invalid.xml'));

    const res = await get(
      'org.example:package',
      'https://invalid_metadata_repo',
      baseUrl
    );

    expect(res).toMatchSnapshot();
  });

  it('skips registry with invalid XML', async () => {
    mockGenericPackage({ html: null });
    httpMock
      .scope('https://invalid_metadata_repo')
      .get('/org/example/package/maven-metadata.xml')
      .reply(200, '###');

    const res = await get(
      'org.example:package',
      'https://invalid_metadata_repo',
      baseUrl
    );

    expect(res).toMatchSnapshot();
  });

  it('handles optional slash at the end of registry url', async () => {
    mockGenericPackage({ html: null });
    const resA = await get('org.example:package', baseUrl.replace(/\/+$/, ''));
    mockGenericPackage({ html: null });
    const resB = await get('org.example:package', baseUrl.replace(/\/*$/, '/'));
    expect(resA).not.toBeNull();
    expect(resB).not.toBeNull();
    expect(resA?.releases).toEqual(resB?.releases);
  });

  it('returns null for invalid registryUrls', async () => {
    const res = await get(
      'org.example:package',

      '${project.baseUri}../../repository/'
    );
    expect(res).toBeNull();
  });

  it('supports scm.url values prefixed with "scm:"', async () => {
    const pom = Fixtures.get('pom.scm-prefix.xml');
    mockGenericPackage({ pom, html: null });

    const res = await get();

    expect(res?.sourceUrl).toBe('https://github.com/example/test');
  });

  it('removes authentication header after redirect', async () => {
    process.env.RENOVATE_EXPERIMENTAL_NO_MAVEN_POM_CHECK = 'true';

    const frontendHost = 'frontend_for_private_s3_repository';
    const frontendUrl = `https://${frontendHost}/maven2`;
    const backendUrl = 'https://private_s3_repository/maven2';
    const metadataPath = '/org/example/package/maven-metadata.xml';
    const pomfilePath = '/org/example/package/2.0.0/package-2.0.0.pom';
    const queryStr = '?X-Amz-Algorithm=AWS4-HMAC-SHA256';

    hostRules.add({
      hostType: datasource,
      matchHost: frontendHost,
      username: 'username',
      password: 'password',
      timeout: 20000,
    });

    httpMock
      .scope(frontendUrl)
      .get(metadataPath)
      .basicAuth({ user: 'username', pass: 'password' })
      .reply(302, '', {
        Location: `${backendUrl}${metadataPath}${queryStr}`,
      })
      .get(pomfilePath)
      .basicAuth({ user: 'username', pass: 'password' })
      .reply(302, '', {
        Location: `${backendUrl}${pomfilePath}${queryStr}`,
      });
    httpMock
      .scope(backendUrl, { badheaders: ['authorization'] })
      .get(`${metadataPath}${queryStr}`)
      .reply(200, Fixtures.get('metadata.xml'))
      .get(`${pomfilePath}${queryStr}`)
      .reply(200, Fixtures.get('pom.xml'));

    const res = await get('org.example:package', frontendUrl);

    expect(res).toMatchSnapshot();
  });

  describe('fetching parent info', () => {
    const parentPackage = {
      dep: 'org.example:parent',
      meta: null,
      pom: Fixtures.get('parent-scm-homepage/pom.xml'),
      latest: '1.0.0',
      jars: null,
      snapshots: [],
    };

    it('should get source and homepage from parent', async () => {
      mockGenericPackage({
        meta: Fixtures.get('child-no-info/meta.xml'),
        pom: Fixtures.get('child-no-info/pom.xml'),
        latest: '2.0.0',
        jars: { '2.0.0': 200 },
        snapshots: [],
        html: null,
      });
      mockGenericPackage(parentPackage);

      const res = await get();

      expect(res).toMatchObject({
        sourceUrl: 'https://github.com/parent-scm/parent',
        homepage: 'https://parent-home.example.com',
      });
    });

    it('should deal with missing parent fields', async () => {
      mockGenericPackage({
        meta: Fixtures.get('child-empty/meta.xml'),
        pom: Fixtures.get('child-empty/pom.xml'),
        latest: '2.0.0',
        jars: { '2.0.0': 200 },
        snapshots: [],
        html: null,
      });

      const res = await get();

      expect(res).toMatchObject({
        display: 'org.example:package',
        group: 'org.example',
        name: 'package',
      });
      expect(res).not.toHaveProperty('homepage');
      expect(res).not.toHaveProperty('sourceUrl');
    });

    it('should deal with circular hierarchy', async () => {
      const parentPom = Fixtures.get('child-parent-cycle/parent.pom.xml');
      const parentPomMock = {
        dep: 'org.example:parent',
        meta: null,
        pom: parentPom,
        latest: '2.0.0',
        jars: null,
        snapshots: [],
      };

      const childMeta = Fixtures.get('child-parent-cycle/child.meta.xml');
      const childPom = Fixtures.get('child-parent-cycle/child.pom.xml');
      const childPomMock = {
        dep: 'org.example:child',
        meta: null,
        pom: childPom,
        latest: '2.0.0',
        jars: null,
        snapshots: [],
      };

      mockGenericPackage({
        ...childPomMock,
        meta: childMeta,
        jars: { '2.0.0': 200 },
        snapshots: [],
        html: null,
      });
      mockGenericPackage(parentPomMock);
      mockGenericPackage(childPomMock);
      mockGenericPackage(parentPomMock);
      mockGenericPackage(childPomMock);
      mockGenericPackage(parentPomMock);

      const res = await get('org.example:child');

      expect(res).toMatchObject({
        homepage: 'https://parent-home.example.com',
      });
    });

    it('should get source from own pom and homepage from parent', async () => {
      mockGenericPackage({
        meta: Fixtures.get('child-scm/meta.xml'),
        pom: Fixtures.get('child-scm/pom.xml'),
        latest: '2.0.0',
        jars: { '2.0.0': 200 },
        snapshots: [],
        html: null,
      });
      mockGenericPackage(parentPackage);

      const res = await get();

      expect(res).toMatchObject({
        sourceUrl: 'https://github.com/child-scm/child',
        homepage: 'https://parent-home.example.com',
      });
    });

    it('should get homepage from own pom and source from parent', async () => {
      mockGenericPackage({
        meta: Fixtures.get('child-url/meta.xml'),
        pom: Fixtures.get('child-url/pom.xml'),
        latest: '2.0.0',
        jars: { '2.0.0': 200 },
        snapshots: [],
        html: null,
      });
      mockGenericPackage(parentPackage);

      const res = await get();

      expect(res).toMatchObject({
        sourceUrl: 'https://github.com/parent-scm/parent',
        homepage: 'https://child-home.example.com',
      });
    });

    it('should get homepage and source from own pom', async () => {
      mockGenericPackage({
        meta: Fixtures.get('child-all-info/meta.xml'),
        pom: Fixtures.get('child-all-info/pom.xml'),
        latest: '2.0.0',
        jars: { '2.0.0': 200 },
        snapshots: [],
        html: null,
      });

      const res = await get();

      expect(res).toMatchObject({
        sourceUrl: 'https://github.com/child-scm/child',
        homepage: 'https://child-home.example.com',
      });
    });

    it('should be able to detect git@github.com:child-scm as valid sourceUrl', async () => {
      mockGenericPackage({
        meta: Fixtures.get('child-scm-gitatcolon/meta.xml'),
        pom: Fixtures.get('child-scm-gitatcolon/pom.xml'),
        latest: '2.0.0',
        jars: { '2.0.0': 200 },
        snapshots: [],
        html: null,
      });

      const res = await get();

      expect(res).toMatchObject({
        sourceUrl: 'https://github.com/child-scm/child',
      });
    });

    it('should be able to detect git@github.com/child-scm as valid sourceUrl', async () => {
      mockGenericPackage({
        meta: Fixtures.get('child-scm-gitatslash/meta.xml'),
        pom: Fixtures.get('child-scm-gitatslash/pom.xml'),
        latest: '2.0.0',
        jars: { '2.0.0': 200 },
        snapshots: [],
        html: null,
      });

      const res = await get();

      expect(res).toMatchObject({
        sourceUrl: 'https://github.com/child-scm/child',
      });
    });

    it('should be able to detect git://@github.com/child-scm as valid sourceUrl', async () => {
      mockGenericPackage({
        meta: Fixtures.get('child-scm-gitprotocol/meta.xml'),
        pom: Fixtures.get('child-scm-gitprotocol/pom.xml'),
        latest: '2.0.0',
        jars: { '2.0.0': 200 },
        snapshots: [],
        html: null,
      });

      const res = await get();

      expect(res).toMatchObject({
        sourceUrl: 'https://github.com/child-scm/child',
      });
    });
  });
});
