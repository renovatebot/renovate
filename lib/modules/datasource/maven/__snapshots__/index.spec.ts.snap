// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/datasource/maven/index falls back to next registry url 1`] = `
{
  "display": "org.example:package",
  "group": "org.example",
  "homepage": "https://package.example.org/about",
  "name": "package",
  "registryUrl": "https://repo.maven.apache.org/maven2",
  "releases": [
    {
      "version": "0.0.1",
    },
    {
      "releaseTimestamp": "2020-01-01T01:00:00.000Z",
      "version": "1.0.0",
    },
    {
      "releaseTimestamp": "2020-01-01T01:00:03.000Z",
      "version": "1.0.3-SNAPSHOT",
    },
    {
      "releaseTimestamp": "2020-01-01T02:00:00.000Z",
      "version": "2.0.0",
    },
  ],
}
`;

exports[`modules/datasource/maven/index ignores unsupported protocols 1`] = `
[
  {
    "version": "0.0.1",
  },
  {
    "releaseTimestamp": "2020-01-01T01:00:00.000Z",
    "version": "1.0.0",
  },
  {
    "releaseTimestamp": "2020-01-01T01:00:03.000Z",
    "version": "1.0.3-SNAPSHOT",
  },
  {
    "releaseTimestamp": "2020-01-01T02:00:00.000Z",
    "version": "2.0.0",
  },
]
`;

exports[`modules/datasource/maven/index removes authentication header after redirect 1`] = `
{
  "display": "org.example:package",
  "group": "org.example",
  "homepage": "https://package.example.org/about",
  "name": "package",
  "registryUrl": "https://frontend_for_private_s3_repository/maven2",
  "releases": [
    {
      "version": "0.0.1",
    },
    {
      "version": "1.0.0",
    },
    {
      "version": "1.0.1",
    },
    {
      "version": "1.0.2",
    },
    {
      "version": "1.0.3-SNAPSHOT",
    },
    {
      "version": "1.0.4-SNAPSHOT",
    },
    {
      "version": "1.0.5-SNAPSHOT",
    },
    {
      "version": "2.0.0",
    },
  ],
}
`;

exports[`modules/datasource/maven/index returns releases 1`] = `
{
  "display": "org.example:package",
  "group": "org.example",
  "homepage": "https://package.example.org/about",
  "name": "package",
  "registryUrl": "https://repo.maven.apache.org/maven2",
  "releases": [
    {
      "version": "0.0.1",
    },
    {
      "releaseTimestamp": "2020-01-01T01:00:00.000Z",
      "version": "1.0.0",
    },
    {
      "releaseTimestamp": "2020-01-01T01:00:03.000Z",
      "version": "1.0.3-SNAPSHOT",
    },
    {
      "releaseTimestamp": "2020-01-01T02:00:00.000Z",
      "version": "2.0.0",
    },
  ],
}
`;

exports[`modules/datasource/maven/index returns releases from custom repository 1`] = `
{
  "display": "org.example:package",
  "group": "org.example",
  "homepage": "https://package.example.org/about",
  "name": "package",
  "registryUrl": "https://custom.registry.renovatebot.com",
  "releases": [
    {
      "version": "0.0.1",
    },
    {
      "releaseTimestamp": "2020-01-01T01:00:00.000Z",
      "version": "1.0.0",
    },
    {
      "releaseTimestamp": "2020-01-01T01:00:03.000Z",
      "version": "1.0.3-SNAPSHOT",
    },
    {
      "releaseTimestamp": "2020-01-01T02:00:00.000Z",
      "version": "2.0.0",
    },
  ],
}
`;

exports[`modules/datasource/maven/index skips registry with invalid XML 1`] = `
{
  "display": "org.example:package",
  "group": "org.example",
  "homepage": "https://package.example.org/about",
  "name": "package",
  "registryUrl": "https://repo.maven.apache.org/maven2",
  "releases": [
    {
      "version": "0.0.1",
    },
    {
      "releaseTimestamp": "2020-01-01T01:00:00.000Z",
      "version": "1.0.0",
    },
    {
      "releaseTimestamp": "2020-01-01T01:00:03.000Z",
      "version": "1.0.3-SNAPSHOT",
    },
    {
      "releaseTimestamp": "2020-01-01T02:00:00.000Z",
      "version": "2.0.0",
    },
  ],
}
`;

exports[`modules/datasource/maven/index skips registry with invalid metadata structure 1`] = `
{
  "display": "org.example:package",
  "group": "org.example",
  "homepage": "https://package.example.org/about",
  "name": "package",
  "registryUrl": "https://repo.maven.apache.org/maven2",
  "releases": [
    {
      "version": "0.0.1",
    },
    {
      "releaseTimestamp": "2020-01-01T01:00:00.000Z",
      "version": "1.0.0",
    },
    {
      "releaseTimestamp": "2020-01-01T01:00:03.000Z",
      "version": "1.0.3-SNAPSHOT",
    },
    {
      "releaseTimestamp": "2020-01-01T02:00:00.000Z",
      "version": "2.0.0",
    },
  ],
}
`;
